docker/backup.mssql.plugin

```diff
@@ -0,0 +1,210 @@
+#!/bin/bash
+# =================================================================================================================
+# MSSQL Backup and Restore Functions:
+# - Dynamically loaded as a plug-in
+# -----------------------------------------------------------------------------------------------------------------
+export serverDataDirectory="/var/opt/mssql/data"
+
+function onBackupDatabase(){
+  (
+    local OPTIND
+    local unset flags
+    while getopts : FLAG; do
+      case $FLAG in
+        ? ) flags+="-${OPTARG} ";;
+      esac
+    done
+    shift $((OPTIND-1))
+
+    _databaseSpec=${1}
+    _backupFile=${2}
+
+    _hostname=$(getHostname ${_databaseSpec})
+    _database=$(getDatabaseName ${_databaseSpec})
+    _port=$(getPort ${_databaseSpec})
+    _portArg=${_port:+"--port=${_port}"}
+    _username=$(getUsername ${_databaseSpec})
+    _password=$(getPassword ${_databaseSpec})
+    echoGreen "Backing up '${_hostname}${_port:+:${_port}}${_database:+/${_database}}' to '${_backupFile}' ..."
+
+    #TODO: add support for backing up transaction log as well.
+    sqlcmd -S ${_hostname} -U ${_username} -P ${_password} -Q "BACKUP DATABASE ${_database} TO DISK = N'${_fileName}' WITH NOFORMAT, NOINIT, SKIP, NOREWIND, NOUNLOAD, STATS = 10"
+    return ${?}
+  )
+}
+
+function onRestoreDatabase(){
+  (
+    local OPTIND
+    local unset flags
+    while getopts : FLAG; do
+      case $FLAG in
+        ? ) flags+="-${OPTARG} ";;
+      esac
+    done
+    shift $((OPTIND-1))
+
+    _databaseSpec=${1}
+    _fileName=${2}
+    _adminPassword=${3}
+
+    _hostname=$(getHostname ${flags} ${_databaseSpec})
+    _database=$(getDatabaseName ${_databaseSpec})
+    _port=$(getPort ${flags} ${_databaseSpec})
+    _portArg=${_port:+"--port=${_port}"}
+    _username=$(getUsername ${_databaseSpec})
+    _password=$(getPassword ${_databaseSpec})
+    echo -e "Restoring '${_fileName}' to '${_hostname}${_port:+:${_port}}${_database:+/${_database}}' ...\n" >&2
+
+    #force single user mode on database to ensure restore works properly
+    sqlcmd -S ${_hostname} -U ${_username} -P ${_password} -Q "ALTER DATABASE ${_database} SET SINGLE_USER WITH ROLLBACK AFTER 30;RESTORE DATABASE ${_database} FROM DISK = N'${_fileName}' WITH FILE = 1, NOUNLOAD, REPLACE, STATS=5;ALTER DATABASE ${_database} SET MULTI_USER"
+    return ${?}
+  )
+}
+
+function onStartServer(){
+  (
+    local OPTIND
+    local unset flags
+    while getopts : FLAG; do
+      case $FLAG in
+        ? ) flags+="-${OPTARG} ";;
+      esac
+    done
+    shift $((OPTIND-1))
+
+    _databaseSpec=${1}
+
+    /opt/mssql/bin/sqlservr --accept-eula &
+  )
+}
+
+function onStopServer(){
+  (
+    local OPTIND
+    local unset flags
+    while getopts : FLAG; do
+      case $FLAG in
+        ? ) flags+="-${OPTARG} ";;
+      esac
+    done
+    shift $((OPTIND-1))
+
+    pkill -f sqlservr
```

Comment:
Is there a gentler and cleaner way of shutdown the server using a management  command?

Reply:
I'll check. sqlcmd doesn't have a nice startup command but shutdown may be an option. pkill doesn't seem to force kill it, from the looks of it a termination notification is sent and shutdown occurs within 30 seconds.

Reply:
oook. So sqlcmd does have a shutdown command but that would require login credentials, which aren't passed to this function. Would be possible using 'sa:MSSQL_SA_PASSWORD' credentials since that env var is required for server startup, but that seems messy.

Not sure if this is relevant but the ouput from both commands is identical:
pkill -f sqlservr
pkill -f sqlservr
sh-4.4$ 2020-05-19 19:35:34.73 spid13s     SQL Server is terminating in response to a 'stop' request from Service Control Manager. This is an informational message only. No user action is required.
2020-05-19 19:35:34.84 spid33s     Service Broker manager has shut down.
2020-05-19 19:35:34.84 spid13s     .NET Framework runtime has been stopped.
2020-05-19 19:35:34.92 spid13s     SQL Trace was stopped due to server shutdown. Trace ID = '1'. This is an informational message only; no user action is required.

sqlcmd 
2020-05-19 19:38:20.65 spid33s     Service Broker manager has shut down.
2020-05-19 19:38:20.65 spid51      .NET Framework runtime has been stopped.
2020-05-19 19:38:20.71 spid51      Server shut down by request from login sa.
2020-05-19 19:38:20.71 spid51      SQL Trace was stopped due to server shutdown. Trace ID = '1'. This is an informational message only; no user action is required.